ZachG update parametric documentation and add contributing doc (#3325)
DataDog/system-tests: docs/scenarios/parametric_contributing.md

# Contributing to Parametric System-tests

Note: a more in-depth overview of parametric system-tests can be found in [parametric.md](parametric.md).

**MUST:** Acquaint yourself with [this section](parametric.md#architecture-how-system-tests-work) for reference so you understand/can track what system-tests are actually doing.

## Use cases

> Let's figure out if your feature is a good candidate to be tested with parametric system-tests. Parametric system-tests are great for assuring uniform behavior between tracers e.g. [environment variable configuration effects on api methods, sampling, propagation, configuration, telemetry](/tests/parametric).

Reviewer: I think it's quite important to emphase here the fact that "parametric" stands for parameters. The original purpose of parametric scenario us when a behavior must be tested across several different value of a parameter. 

Using end-to-end scenarios for this use case is a pain, as we need to spawn an infra for each parameter value. But parametric scenario is way better, as it starts on container per test case, and it's reasonnably fast.

But if your usage does not implies several parameters, then an end-to-end scenario **must** be used : it'll achieve the same level of behavior uniformity verification. And it'll test the feature on real world use case, catching way more issues.

I think making this point very clear is very important, as parametric scenario has been a little bit abused by the past. And is it does not scale very well, if we can keep it for the relevant use case, it'll be for the best of all users.

Author: Thanks for this explanation! I'll add pieces of it in!

The parametric tests rely on the hitting of [http endpoints](/tests/parametric) that run tracer methods to produce and modify spans (manual instrumentation). If you'd like to test behavior across automatic instrumentations of tracers then you should assess if weblog system-tests may be a better fit.

Parametric system-tests are horrible for testing internal or niche tracer behavior. Tests for those should exist on the tracer repos since they're only applicable for that specific tracer.

## Getting setup

> Usually the one writing the system-tests is writing for a new feature, potentially one that hasn't been completed across all tracers yet. Therefore they'll want to focus on writing and getting the tests to pass for their tracer implementation first.

Reviewer 2: ```suggestion
We usually add new system tests when validating a new feature. This feature might not yet be implemented across all dd-trace libraries. If at least one library already supports the feature, you can verify your test by running it against that library
```

To begin we need to point system-tests towards a tracer that has the feature implemented (published or on a branch).

> Follow [Binaries Documentation](../execute/binaries.md) for your particular tracer language to set this up.

Reviewer 2: Well, we will "point" system-tests toward the library when we run the test with TEST_LIBRARY=x. But if the library contains the feature in an unmerged branch, then we need to follow binaries.md

Author: Gotcha, you're interpreting "point" here as the language choice, I meant as the actual tracer. I think that parametric.md#running-the-tests takes care of talking about how to run the test for a specific language. I'll make it clearer though here.

Author: ```suggestion
To begin we need to make sure system-tests run with a tracer that has implemented the feature being tested (published or on a branch).
Follow [Binaries Documentation](../execute/binaries.md) for your particular tracer language to set this up.
```

> [Try running the tests for your tracer language](parametric.md#running-the-tests) and make sure some pass (no need to run the whole suite, you can stop the tests from running with `ctrl+c`). If you have an issue, checkout the [debugging section](parametric.md#debugging) to troubleshoot.

Reviewer 2: This is kinda vague, what do you mean?
Like in the case where your tracer language is not the "sample working library"?

Author: I just want implementers to try running some of the test suite with their custom tracer and making sure it works before they try to start actually writing tests. Often times there are issues with just running the tests at all. This essentially a verification step. I'll make that clearer.

Author: ```suggestion
[Verify that you can run some parametric tests with your custom tracer](parametric.md#running-the-tests). Make sure some pass (no need to run the whole suite, you can stop the tests from running with `ctrl+c`). If you have an issue, checkout the [debugging section](parametric.md#debugging) to troubleshoot.
```

## Writing the tests

Now that we're all setup with a working test suite and a tracer with the implemented feature, we can begin writing the new tests.

> First take a look at the [currently existing tests](/tests/parametric) and see if what you're trying to test is similar and can use the same methods/endpoints (in many cases this is true).

Reviewer 2: "methods/endpoints" <- should we link to http server implementation details here?

Author: I clarify and link in the following two paragraphs.

> For all of the exact methods already implemented you can take a look at `class APMLibrary` in the [_library_client.py](/utils/parametric/_library_client.py). If you're wondering exactly what the methods do, you can take at look at the respective endpoints they're calling in that same file in `class APMLibraryClient`.

Reviewer 2: ```suggestion
For a list of methods that already exist, refer to `class APMLibrary` in the [_library_client.py](/utils/parametric/_library_client.py). If you're wondering what the methods do, you can take at look at the respective endpoints they're calling in that same file in `class APMLibraryClient`.
```

> The endpoints (where the actual tracer code runs) are defined in the Http Server implementations per tracer [listed here](parametric.md#http-server-implementations). Click on the one for your language to take a look at the endpoints. In some cases you may need to just slightly modify an endpoint rather than add a new one.

Reviewer 2: I feel it would be helpful to define "Parametric servers" as the servers that serve requests (via endpoints) from APMLibraryClient. They perform tracing activity and send relevant data about the resulting trace(s) back to the client.

Author: I feel like we should rely on the documentation in parametric.md for this understanding, which is why I link to it extensively and ask users to read the section of how system-tests work before starting in on this doc. I'll add a reminder to reference that section if confused.

### If you need to add additional endpoints to test your new feature

*Note:* please refer to the [architecture section](parametric.md#architecture-how-system-tests-work) if you're confused throughout this process.

Then we need to do the following:

> * Determine what you want the endpoint to be called and what you need it to do, and add it to your tracer's http server.

Reviewer 2: Good opportunity to use "parametric server" or "parametric app" instead - 
```suggestion
* Determine what you want the endpoint to be called and what you need it to do, and add it to your tracer's parametric app.
```

Author: Hmmm I see both sides. It feels like we call them http servers in parametric.md right? I don't want to change the naming. Let me add a bit more explanation and see if it seems clearer.

*Note:* If adding a new endpoint please let a Python implementer know so they can add it as well [see](parametric.md#shared-interface)
* In [_library_client.py](/utils/parametric/_library_client.py) Add both the endpoint call in `class APMLibraryClient` and the method that invokes it in `class APMLibrary`. Use other implementations for reference.
* Ok we now have our new method! Use it in the tests you write using the [below section](#if-the-methods-you-need-to-run-your-tests-are-already-written)

### If the methods you need to run your tests are already written

> Make a new test file in `tests/parametric`, copying in the testing code you want to use as a base/guideline (usually the class and and one of the test methods in it).

Reviewer 2: Don't always need to create a new file 🤷

Author: Good point!

Author: ```suggestion
If it makes sense to add your tests to a file that already exists, great! Otherwise make a new test file in `tests/parametric`. 

Next copy the testing code you want to use as a base/guideline (usually the class (if using a new file) and one of the test methods in it).
```

Then:

* [Change the name of the feature annotation it'll fit under for the feature parity board](/docs/edit/features.md) (Not always needed e.g. `@features.datadog_headers_propagation` is used for all the propagation features)
* Change the class and method name to fit what you're testing.

> * [Change your tracer's respective manifest.yml file](/docs/edit/manifest.md) or else the script won't know to run your new test. If you're confused at how to do this properly, search for the file you copied the test from in the manifest file and see how it's specified, you can probably copy that for your new file (make sure the path is the same).

Reviewer 2: "Change" doesn't seem like the right word choice here; maybe "Choose"?

Author: I get your point but I think that in this case "change" makes more sense since we're asking them to copy code before. "Copy that method, ok now change the method name" Seems like more common speak and direct for us than "Copy that method, ok now choose a method name". The former would definitely be clearer to me if I were pair programming and driving.

Author: Feel free to reopen if you feel strongly.

# Contributing to Parametric System-tests

Note: a more in-depth overview of parametric system-tests can be found in [parametric.md](parametric.md).

**MUST:** Acquaint yourself with [how system tests work](parametric.md#architecture-how-system-tests-work) before proceeding.

## Use cases

Let's figure out if your feature is a good candidate to be tested with parametric system-tests. 

Parametric system-tests are great for assuring uniform behavior between tracers e.g. [environment variable configuration effects on api methods, sampling, propagation, configuration, telemetry](/tests/parametric).

Parametric tests make requests to [http endpoints](/tests/parametric) dedicated to various tracer methods for creating and modifying spans (manual instrumentation). If you want to test automatic instrumentation behavior, weblog system-tests may be a better fit.

System-tests are **not** for testing internal or niche tracer behavior. Unit tests are a better fit for that case.

## Getting set up

> We usually add new system tests when validating a new feature. This feature might not yet be implemented across all dd-trace libraries. If at least one library already supports the feature, you can verify your test by running it against that library

Author: ```suggestion
We usually add new system tests when validating a new feature. This feature might not yet be implemented across all dd-trace libraries. If at least one library already supports the feature, you can verify your test by running it against that library.
```

# Contributing to Parametric System-tests

Note: a more in-depth overview of parametric system-tests can be found in [parametric.md](parametric.md).

**MUST:** Acquaint yourself with [how system tests work](parametric.md#architecture-how-system-tests-work) before proceeding.

## Use cases

Let's figure out if your feature is a good candidate to be tested with parametric system-tests.

System-tests in general are great for assuring uniform behavior between different dd-trace repos (tracing, ASM, DI, profiling, etc.). There are two types of system-tests, [end-to-end](/docs/README.md) and [parametric](/docs/scenarios/parametric.md).

> Parametric tests in the Datadog system test repository validate the behavior of APM Client Libraries by interacting only with their public interfaces. These tests ensure the telemetry generated (spans, metrics, instrumentation telemetry) is consistent and accurate when libraries handle different input parameters (e.g., calling a Tracer's startSpan method with a specific type) and configurations (e.g., sampling rates, distributed tracing, remote settings). They run against web applications in languages like Java, Go, Python, PHP, Node.js, C++, and .NET, which expose endpoints simulating real-world library usage. The generated telemetry is sent to a Datadog agent, queried, and verified by system tests to confirm proper library functionality across scenarios.

Reviewer 3: ```suggestion
Parametric tests in the Datadog system test repository validate the behavior of APM Client Libraries by interacting only with their public interfaces. These tests ensure the telemetry generated (spans, metrics, instrumentation telemetry) is consistent and accurate when libraries handle different input parameters (e.g., calling a Tracer's startSpan method with a specific type) and configurations (e.g., sampling rates, distributed tracing header formats, remote settings). They run against web applications written in Ruby, Java, Go, Python, PHP, Node.js, C++, and .NET, which expose endpoints simulating real-world ddtrace usage. The generated telemetry is sent to a Datadog agent, queried, and verified by system tests to confirm proper library functionality across scenarios.
```

# Contributing to Parametric System-tests

Note: a more in-depth overview of parametric system-tests can be found in [parametric.md](parametric.md).

**MUST:** Acquaint yourself with [how system tests work](parametric.md#architecture-how-system-tests-work) before proceeding.

## Use cases

Let's figure out if your feature is a good candidate to be tested with parametric system-tests.

> System-tests in general are great for assuring uniform behavior between tracers. There are two types of system-tests, [end-to-end](/docs/README.md) and [parametric](/docs/scenarios/parametric.md).

Reviewer 3: nit: profiling/asm/di also use system tests

```suggestion
System-tests in general are great for assuring uniform behavior between client libraries. There are two types of system-tests, [end-to-end](/docs/README.md) and [parametric](/docs/scenarios/parametric.md).
```

Author: ```suggestion
System-tests in general are great for assuring uniform behavior between different dd-trace repos (tracing, ASM, DI, profiling, etc.). There are two types of system-tests, [end-to-end](/docs/README.md) and [parametric](/docs/scenarios/parametric.md).
```

> The "parametric" in parametric system-tests stands for parameters. The original purpose of parametric scenarios is when a behavior must be tested across several different values for one or more parameters, usually different tracer configurations with some examples being [environment variable configuration effects on api methods, sampling, propagation, configuration, telemetry](/tests/parametric).

Reviewer 3: Here's some information I want to include. I am not sure if it fits in this section but this should be one the first things a contributor sees. It's a quick summary of what parametric tests should be used for:

```suggestion
Parametric tests in the Datadog system test repository validate the behavior of APM Client Libraries by interacting only with their public interfaces. These tests ensure the telemetry generated (spans, metrics, instrumentation telemetry) is consistent and accurate when libraries handle different input parameters (e.g., calling a Tracer's startSpan method with a specific type) and configurations (e.g., sampling rates, distributed tracing, remote settings). They run against web applications in languages like Java, Go, Python, PHP, Node.js, C++, and .NET, which expose endpoints simulating real-world library usage. The generated telemetry is sent to a Datadog agent, queried, and verified by system tests to confirm proper library functionality across scenarios.
```

> If your usage does not require different parameter values, then [end-to-end system-tests](/docs/README.md) should be used as they will achieve the same level of behavior uniformity verification and test the feature on real world use cases, catching more issues.

Author: Oh great write up, I'll link to that.

> If your usage does not require different parameter values, then [end-to-end system-tests](/docs/README.md) should be used as they will achieve the same level of behavior uniformity verification and test the feature on real world use cases, catching more issues.

Author: ```suggestion
If your usage does not require different parameter values, then [end-to-end system-tests](/docs/README.md) should be used as they will achieve the same level of behavior uniformity verification and test the feature on real world use cases, catching more issues. End-to-end tests are also what should be used for verify behavior between tracer integrations.
```

> 

Author: ```suggestion
Fore more on the differences between end-to-end and parametric tests, see [here](/docs/scenarios/README.md#scenarios)
```

System-tests are **not** for testing internal or niche tracer behavior. Unit tests are a better fit for that case.

## Getting set up

We usually add new system tests when validating a new feature. This feature might not yet be implemented across all dd-trace libraries. If at least one library already supports the feature, you can verify your test by running it against that library.

To begin we need to make sure system-tests run with a tracer that has implemented the feature being tested (published or on a branch).
Follow [Binaries Documentation](../execute/binaries.md) for your particular tracer language to set this up.

[Verify that you can run some parametric tests with your custom tracer](parametric.md#running-the-tests). Make sure some pass (no need to run the whole suite, you can stop the tests from running with `ctrl+c`). If you have an issue, checkout the [debugging section](parametric.md#debugging) to troubleshoot.

## Writing the tests

Now that we're all setup with a working test suite and a tracer with the implemented feature, we can begin writing the new tests.

> **MUST:** If you haven't yet, please acquaint yourself with [how system tests work](parametric.md#architecture-how-system-tests-work) before proceeding and reference it throughout this section.

Author: It's purposeful to address other comments you made earlier about explaining more of how system-tests work. Most of that info should go in parametric.md, so rather than essentially re-explain it here I reiterate that implementers should reference it if confused.

First take a look at the [currently existing tests](/tests/parametric) and see if what you're trying to test is similar and can use the same methods/endpoints, in many cases new endpoints do not need to be added.

For a list of client methods that already exist, refer to `class APMLibrary` in the [_library_client.py](/utils/parametric/_library_client.py). If you're wondering what the methods do, you can take at look at the respective endpoints they're calling in that same file in `class APMLibraryClient`.

The endpoints (where the actual tracer code runs) are defined in the Http Server implementations per tracer [listed here](parametric.md#http-server-implementations). Click on the one for your language to take a look at the endpoints. In some cases you may need to just slightly modify an endpoint rather than add a new one.

### If you need to add additional endpoints to test your new feature

> *Note:* please refer to the [architecture section](parametric.md#architecture-how-system-tests-work) if you're confused throughout this process.

Reviewer 3: We should link to this dashboard to show endpoints that exist but are not up to spec: https://feature-parity.us1.prod.dog/#/?runDateFilter=7d&feature=339

Author: I'm going to put this in parametric.md instead.  I don't think this should be in the main path of a first time contributor.

Then we need to do the following:

> * Determine what you want the endpoint to be called and what you need it to do, and add it to your tracer's http server.

Reviewer 3: Only add new endpoints that operate on the public API and execute ONE operation. Endpoints that execute complex operations or validate tracer internals will not be accepted. 

Also parametric tests should not test the behavior of ddtrace integrations. weblog apps should be used for that

*Note:* If adding a new endpoint please let a Python tracer implementer know so they can add it as well [see](parametric.md#shared-interface)

> 

Author: ```suggestion
*Note*: Only add new endpoints that operate on the public API and execute ONE operation. Endpoints that execute complex operations or validate tracer internals will not be accepted.
```
